cmake: add compiler check for ASAN

I need to compile VisualBoyAdvance-M with Network Link, because it's not enabled for some reason in the official build of 2.1.1, however going to compile it complains that the compiler can't find SFML or SSP (gcc-4.8) and GTK (gtk-2.0/gtk-3.0) even though they exist, dev packages and all.

```
/usr/include/SFML
[...]
/usr/share/SFML
/usr/share/SFML/cmake
/usr/share/SFML/cmake/Modules
/usr/share/SFML/cmake/Modules/FindSFML.cmake
/usr/include/gtk-2.0
[...]
/usr/include/gtk-3.0
```


```
$ cmake ..
Could NOT find SFML (missing:  SFML_NETWORK_LIBRARY SFML_SYSTEM_LIBRARY)
CMake Warning at CMakeLists.txt:232 (message):
  SFML not found, LINK will be disabled


-- Found gcc-ar: /usr/bin/gcc-ar
-- Found gcc-nm: /usr/bin/gcc-nm
-- Found gcc-ranlib: /usr/bin/gcc-ranlib
-- Could NOT find SSP (missing: SSP_LIBRARY) 
CMake Error at src/wx/CMakeLists.txt:538 (message):
  Could not find gtk


-- Configuring incomplete, errors occurred!
```

Looks to me that it's not looking in the right places to find them or using the wrong syntax in the code to search for them, the proper directory is `/usr/include/` is where it's suppose to search for the libraries it needs for compiling, and SSP is has since been included in GCC since 4.1. The error message for "Could not find gtk" is too vague of a message, what are you expecting? libgtk-2.0(-dev)? gtk-3.0(-dev)?

[CMakeOutput.log](https://gist.github.com/Alkaris/c46f6ede5ea281cfac6b4e20dfd83db4#file-cmakeoutput-log)
[CMakeError.log](https://gist.github.com/Alkaris/2f47cda1db096e55c48b496babe488ea#file-cmakeerror-log)

```
Linux Mint 19.1 4.15.0-46-generic #49-Ubuntu x86_64 GNU/Linux
```

you can run *installdeps* to get dependencies

On Mon, Apr 8, 2019 at 7:14 PM Alkaris <notifications@github.com> wrote:

> I need to compile VisualBoyAdvance-M with Network Link, because it's not
> enabled for some reason in the official build of 2.1.1, however going to
> compile it complains that the compiler can't find SFML or SSP (gcc-4.8) and
> GTK (gtk-2.0/gtk-3.0) even though they exist, dev packages and all.
>
> /usr/include/SFML
> [...]
> /usr/share/SFML
> /usr/share/SFML/cmake
> /usr/share/SFML/cmake/Modules
> /usr/share/SFML/cmake/Modules/FindSFML.cmake
> /usr/include/gtk-2.0
> [...]
> /usr/include/gtk-3.0
>
> $ cmake ..
> Could NOT find SFML (missing:  SFML_NETWORK_LIBRARY SFML_SYSTEM_LIBRARY)
> CMake Warning at CMakeLists.txt:232 (message):
>   SFML not found, LINK will be disabled
>
>
> -- Found gcc-ar: /usr/bin/gcc-ar
> -- Found gcc-nm: /usr/bin/gcc-nm
> -- Found gcc-ranlib: /usr/bin/gcc-ranlib
> -- Could NOT find SSP (missing: SSP_LIBRARY)
> CMake Error at src/wx/CMakeLists.txt:538 (message):
>   Could not find gtk
>
>
> -- Configuring incomplete, errors occurred!
>
> Looks to me that it's not looking in the right places to find them or
> using the wrong syntax in the code to search for them, the proper directory
> is /usr/include/ is where it's suppose to search for the libraries it
> needs for compiling. The error message for "Could not find gtk" is too
> vague of a message, what are you expecting? libgtk-2.0(-dev)? gtk-3.0(-dev)?
>
> CMakeOutput.log
> <https://gist.github.com/Alkaris/c46f6ede5ea281cfac6b4e20dfd83db4#file-cmakeoutput-log>
> CMakeError.log
> <https://gist.github.com/Alkaris/2f47cda1db096e55c48b496babe488ea#file-cmakeerror-log>
>
> Linux Mint 19.1 4.15.0-46-generic #49-Ubuntu x86_64 GNU/Linux
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/visualboyadvance-m/visualboyadvance-m/issues/409>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AWPDtqelNlvHiw4EbTIF1_mfYqsVvB5uks5veySngaJpZM4ch6zp>
> .
>

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -480,6 +480,8 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID STREQUAL Clang)
         set(MY_C_DBG_FLAGS -g -fno-omit-frame-pointer)
     endif()
 
+    include(CheckCXXCompilerFlag)
+
     if(ENABLE_ASAN)
         string(TOLOWER ${CMAKE_BUILD_TYPE} build)
         if(NOT build STREQUAL "debug")
@@ -490,7 +492,13 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID STREQUAL Clang)
         if(SANITIZER STREQUAL "on" OR SANITIZER STREQUAL "true")
             set(SANITIZER address)
         endif()
-        set(MY_C_DBG_FLAGS ${MY_C_DBG_FLAGS} -fsanitize=${SANITIZER} -lasan)
+        set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} "-lasan")
+        check_cxx_compiler_flag("-fsanitize=${SANITIZER} -lasan" ASAN_SUPPORT_FLAG)
+        if(${ASAN_SUPPORT_FLAG})
+            set(MY_C_DBG_FLAGS ${MY_C_DBG_FLAGS} -fsanitize=${SANITIZER} -lasan)
+        else()
+            message(FATAL_ERROR "asan not available to compiler.")
+        endif()
     endif()
 
     # common flags
@@ -499,8 +507,6 @@ if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID STREQUAL Clang)
     if(NOT (WIN32 OR X86_32)) # inline asm is not allowed with -fPIC
         set(MY_C_FLAGS ${MY_C_FLAGS} -fPIC)
     endif()
-
-    include(CheckCXXCompilerFlag)
 
     # check if ssp flags are supported for this version of gcc
     if(CMAKE_COMPILER_IS_GNUCXX)

